tests/cxx/iwyu_stricter_than_cpp-autocast.h

```diff
@@ -65,51 +63,111 @@ template <typename T>
     const TplIndirectStruct2<char>& ic2) {
 }
 
+// --- With user-defined types as template parameters.
+
+// struct IndirectStruct2; (fwd-declared above)
+template <typename TFullTypeUsed, typename TForwardDeclarable>
+struct TplIndirectStruct3;
+
+void TplFnValues(
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    TplIndirectStruct3<IndirectStruct1, IndirectStruct2>
+        onlyArgumentTypeProvided,
+    TplIndirectStruct3<IndirectStruct2, IndirectStruct2> allForwardDeclared,
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    TplDirectStruct7<IndirectStruct1, IndirectStruct2> allNeededTypesProvided,
+    TplDirectStruct7<IndirectStruct2, IndirectStruct2> onlyTemplateProvided);
+
+void TplFnRefs(
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    const TplIndirectStruct3<IndirectStruct1, IndirectStruct2>&
+        onlyArgumentTypeProvided,
+    const TplIndirectStruct3<IndirectStruct2, IndirectStruct2>&
+        allForwardDeclared,
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    const TplDirectStruct7<IndirectStruct1, IndirectStruct2>&
+        allNeededTypesProvided,
+    const TplDirectStruct7<IndirectStruct2, IndirectStruct2>&
+        onlyTemplateProvided);
+
+inline void HeaderDefinedTplFnRefs(
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    const TplIndirectStruct3<IndirectStruct1, IndirectStruct2>&
+        onlyArgumentTypeProvided,
+    const TplIndirectStruct3<IndirectStruct2, IndirectStruct2>&
+        allForwardDeclared,
+    // IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+    const TplDirectStruct7<IndirectStruct1, IndirectStruct2>&
+        allNeededTypesProvided,
+    const TplDirectStruct7<IndirectStruct2, IndirectStruct2>&
+        onlyTemplateProvided) {
+}
+
 // --- The rules do not apply for friend functions declarations.
 
 struct AutocastStruct {
   // IWYU: IndirectStruct1 needs a declaration
   friend void ClassFn1(const IndirectStruct1&);
   // IWYU: TplIndirectStruct1 needs a declaration
   friend void ClassFn2(TplIndirectStruct1<char>);
+  friend void ClassFn3(
+      // IWYU: IndirectStruct1 needs a declaration
+      const TplIndirectStruct3<IndirectStruct1, IndirectStruct2>&);
 };
 
+// Test that IWYU finds constructors even if they are not instantiated.
+
+template <typename T>
+struct NonInstantiatedCtor {
+  NonInstantiatedCtor(int) {
+  }
+  T t;
+};
+// IWYU: IndirectStruct1 is...*iwyu_stricter_than_cpp-i1.h.*for autocast
+void Fn(NonInstantiatedCtor<IndirectStruct1>);
+template <typename T>
+struct NoAutocastForSpec {
+  NoAutocastForSpec(int) {
+  }
+  T t;
+};
+template <>
+// IWYU: IndirectStruct1 needs a declaration
+struct NoAutocastForSpec<IndirectStruct1> {};
+// IWYU: IndirectStruct1 needs a declaration
+void Fn(NoAutocastForSpec<IndirectStruct1>);
```

Comment:
Should there be a TODO here to match https://github.com/include-what-you-use/include-what-you-use/pull/1313/commits/d847df1e54a0299bc0d9d39ef9048925a5291e12#diff-7ec045b2b5739efe5504c9d31e7e935fe9405be09ae655d067bda89fdd9cf6f9R687?